[1-3단계 - 방탈출 예약 관리] 듀이(이주희) 미션 제출합니다.

ljhee92 opened the pull request:
## 체크 리스트
- [x] 미션의 필수 요구사항을 모두 구현했나요?
- [x] Gradle `test`를 실행했을 때, 모든 테스트가 정상적으로 통과했나요?
- [x] 애플리케이션이 정상적으로 실행되나요?
- [x] [프롤로그](https://prolog.techcourse.co.kr)에 셀프 체크를 작성했나요?
  https://prolog.techcourse.co.kr/studylogs/4162

## 객체지향 생활체조 요구사항을 얼마나 잘 충족했다고 생각하시나요?
### 1~5점 중에서 선택해주세요.

- [ ] 1 (전혀 충족하지 못함)
- [ ] 2
- [ ] 3 (보통)
- [x] 4
- [ ] 5 (완벽하게 충족)

### 선택한 점수의 이유를 적어주세요.
- 책임에 따라 클래스를 분리하려고 노력했습니다.
- 모든 원시값을 포장하지는 않았습니다.

## 웹 개발 경험 체크
🌱 레벨2 시작 이전에, 웹 애플리케이션 개발 경험이 있나요?
- [ ] 아니오. 전혀 경험이 없습니다.
- [ ] 예. 예제 코드를 따라하거나 클론 코딩을 해봤습니다. (ex. 강의를 보고 따라함)
- [x] 예. 웹 애플리케이션을 만들었지만 로컬에서만 실행해봤습니다.
- [ ] 예. 웹 애플리케이션을 만들고 배포해봤습니다. (ex. 클라우드 혹은 리모트 서버에 배포)

## 어떤 부분에 집중하여 리뷰해야 할까요?
안녕하세요 범블비!
백엔드 7기 듀이입니다.
레벨 2 첫 리뷰 잘 부탁드립니다.😊
아래 내용은 페어와 함께 '왜' 이런 코드를 짜게 되었는지 논의하면서 정리한 의견입니다.
리뷰하실 때 참고해주시면 감사드리겠습니다!

### `Controller` 분리
- 뷰를 매핑하는 컨트롤러와 데이터를 전달하는 API 컨트롤러의 책임이 다르다고 판단하여 `ViewController`, `RoomescapeController`로 분리하였습니다.
- `VieController` : View만 매핑하므로 `@Controller` 어노테이션 사용
- `RoomescapeController` : 데이터를 전달하므로 `@RestController` 어노테이션 사용

### `Repository` 분리
- 처음에는 `Controller`에 `List<Reservation>`과 `AtomicLong`을 두고 데이터를 저장했습니다.
- 컨트롤러의 역할은 '뷰와 연결되는 지점'이라고만 생각되어 데이터의 저장을 `ReservationRepository`로 분리했습니다.

### `DTO` 사용
- 예약을 생성할 때, 뷰에서 받아오는 정보는 id가 없었으므로 이름, 날짜, 시간 데이터만 포함된 `CreateReservationDto`를 사용했습니다.
- 전체 예약 리스트나 생성된 예약을 뷰에 반환할 때, 예약 날짜와 시간을 분리하여 뷰가 필요한 데이터만 포함된 `ReservationResponseDto`를 사용했습니다.
- 도메인엔티티 -> DTO로의 변환, DTO -> 도메인엔티티로의 변환은 도메인엔티티 내부에서 하기보다 자주 변경되는 DTO 내부에서 하는 것이 맞다고 판단했습니다.

###  커스텀 예외 사용
- 초기 구현에서는 `ReservationRepository#deleteById` 메서드에서 id로 Reservation을 삭제할 때, id에 해당되는 값이 없을 경우 `RuntimeException`을 던져 어떤 예외가 발생하는지 이유를 알기 어려웠습니다.
- 사용자 입장에서 어떤 예외인지 명확하게 알기 위해 내부에서는 `InvalidReservationException`이라는 커스텀 예외를 던지고, 컨트롤러에서 예외를 잡아 HTTP 상태 코드를 반환하도록 처리했습니다.(`RoomexcapeController#deleteReservation`)

### 📌 코드리뷰 질문 그라운드룰 (지우지 마세요)
✅ 많은 사람과 공유하고 토론해보는 것이 더 좋은 질문에는, 리뷰어가 💬말풍선 이모지를 남겨 토론을 추천할 수 있어요.  
✅ 토론을 추천받은 질문은 `#be-7기-잡담-질답`슬랙채널이나 캠퍼스에서 자유롭게 토론을 이어가보세요!  
✅ 코드에 대한 맥락 없이 리뷰어의 주관적인 의견을 묻거나, 단순히 현업에서는 어떻게 하는지를 묻는 질문은 코드 리뷰 본문에는 어울리지 않아요.  
코드 리뷰는 변경된 코드의 목적, 의도, 품질을 중심으로 논의하는 자리예요.  
✅ 코드와 관련된 질문이 있다면 PR 본문에 적기보다는, 해당 코드를 선택해 코멘트를 남겨주세요.

rladmstn committed:
docs: 기능 명세 작성

rladmstn committed:
feat: 예약 홈 화면 연결

rladmstn committed:
feat: 예약 조회 기능 추가

rladmstn committed:
feat: 예약 조회 기능 추가

ljhee92 committed:
feat: 예약 추가/취소 기능 추가

rladmstn committed:
fix: conflict 해결

rladmstn committed:
refactor: 패키지 분리

ljhee92 committed:
refactor: PathVariable 명시

rladmstn committed:
feat: repository 추가

ljhee92 committed:
fix: 시간 포맷 수정 및 뷰 반환 DTO 추가

ljhee92 committed:
refactor: 뷰 컨트롤러와, 데이터 반환 컨트롤러 분리

rladmstn committed:
feat: 예외 처리 추가

ddaaac commented on src/main/java/roomescape/controller/RoomescapeController.java:
```diff
@@ -0,0 +1,50 @@
+    }
+
+    @GetMapping("/reservations")
+    public ResponseEntity<List<ReservationResponseDto>> getReservations(){
```
`ResponseEntity<List<ReservationResponseDto>>`를 리턴하는 것과 `List<ReservationResponseDto>`을 리턴하는 건 어떻게 다를까요?

ljhee92 replied:
`ResponseEntity<List<ReservationResponseDto>>`를 리턴하는 것과 `List<ReservationResponseDto>`를 리턴하는 것에는 이런 차이점이 있다고 생각합니다.
1. 직렬화 처리 방식
 - `List<ReservationResponseDto>`를 리턴하려면 클래스 레벨에 `@RestController` 어노테이션이 있거나 메서드에 `@ResponseBody` 어노테이션이 있어야 스프링이 해당 객체를 JSON으로 직렬화하여 HTTP 응답 본문으로 리턴할 수 있습니다.
 - `ResponseEntity<List<ReservationResponseDto>>`를 리턴할 때에는 `@RestController`, `@ResponseBody`가 없어도 HTTP 응답을 할 수 있는 객체(JSON 등)로 제공됩니다.    

2. HTTP 응답 제어
- `ResponseEntity<List<ReservationResponseDto>>`를 리턴한다면 클라이언트에게 응답을 반환할 때, 응답 본문 외에도 HTTP 상태 코드, 상태 메시지, 응답 헤더를 유연하게 지정하여 응답을 명시적으로 제어할 수 있습니다.

추가로, `ResponseEntity`를 사용하면`@ResponseBody`나 `@RestController`가 없어도 동작은 하지만, 현재 구현한 Controller는 View가 아닌 데이터만 응답하는 Controller임을 명확하게 하기 위해 `@RestController`를 함께 명시해주었습니다.

ddaaac replied:
너무 꼼꼼하게 잘 찾아주셨어요 👍

ddaaac commented on src/main/java/roomescape/controller/RoomescapeController.java:
```diff
@@ -0,0 +1,50 @@
+    }
+
+    @GetMapping("/reservations")
+    public ResponseEntity<List<ReservationResponseDto>> getReservations(){
```
`ResponseEntity<List<ReservationResponseDto>>`를 리턴하는 것과 `ResponseEntity<List<Reservation>>`을 리턴하는 건 어떻게 다를까요?

ljhee92 replied:
`ResponseEntity<List<ReservationResponseDto>>`를 리턴하는 것과 `ResponseEntity<List<Reservation>>`을 리턴하는 것은 이런 차이점이 있다고 생각합니다.
1. 도메인 보호
 - `Reservation` 도메인 객체가 외부로 드러나지 않아 보호할 수 있습니다. 외부에 도메인이 드러날 경우 내부 구현이 변경될 위험이 있습니다.
2. 클라이언트의 요구사항 반영 가능
 - 현재 `Reservation`은 날짜와 시간을 함께 관리하고 있는데, 뷰에서는 날짜와 시간을 분리하여 표기하고 있습니다. 또, 시간의 형식에서 초가 필요하지 않습니다. 이처럼 클라이언트의 요구사항을 유연하게 반영할 수 있습니다.

ddaaac commented on src/main/java/roomescape/repository/ReservationRepository.java:
```diff
@@ -0,0 +1,35 @@
+
+@Repository
+public class ReservationRepository {
+    private final List<Reservation> reservations = new ArrayList<>();
+    private final AtomicLong index = new AtomicLong(1);
```
왜 Long이 아닌 AtomicLong을 사용했을까요?
그 이유를 Reservations에도 동일하게 적용할 수 있을까요?

ljhee92 replied:
사실 처음에는 학습 테스트에 `AtomicLong`이 쓰여 생각 없이 작성했습니다😅
그래서 `AtomicLong`에 대해 찾아보니 `Long`과 다른 차이점이 몇 개 있더라구요!
- 원자성 보장 가능 -> thread safe합니다. 즉, 방탈출 예약이 여러 건 들어올 때 id를 순서대로 지정해줄 수 있습니다.
- 자바에서 index++처럼 값을 증가시키지 않고, `getAndIncrement()` 메서드로 증가시켜줄 수 있어 명시적 표현이 가능합니다.

위와 같은 이유를 `Reservations`에도 동일하게 적용한다면 `CopyOnWriteArrayList`와 `ConcurrentHashMap`을 사용할 수 있을 것 같은데요.
`CopyOnWriteArrayList`의 경우에는 읽기 작업이 빈번할 때 유용하지만 쓰기 작업 시 매 번 전체 배열을 복사한다는 단점이 있습니다. 방탈출 예약 같은 경우 예약(쓰기 작업)이 빈번하게 일어나므로 `CopyOnWriteArrayList`는 적합하지 않다고 생각해 `ConcurrentHashMap`으로 변경했습니다.

[refactor: 동시성 보장을 위해 List -> ConcurrentMap 변경](https://github.com/woowacourse/spring-roomescape-admin/pull/221/commits/73e0a6b1daa455d8bb26a7be8808d711c25d0962)

ddaaac replied:
다음 단계에 DB 적용할 때도 동시성에 대한 고민을 남겨둔 채 가져가보면 좋을 것 같습니다!

ddaaac commented on src/main/java/roomescape/controller/ViewController.java:
```diff
@@ -0,0 +1,18 @@
+import org.springframework.stereotype.Controller;
+import org.springframework.web.bind.annotation.GetMapping;
+
+@Controller
```
`@RestController`와 `@Controller`는 어떤 차이가 있을까요?

ljhee92 replied:
`@Controller`는 뷰를 반환할 수도 있고, 데이터를 반환해줄 수도 있습니다.
데이터를 반환해줄 때에는 `@ResponseBody`를 사용해야 객체를 HTTP 응답 본문으로 직렬화할 수 있습니다.
뷰를 반환하는 전통적인 서버 사이드 렌더링 방식보다 데이터를 반환하는 클라이언트 사이드 렌더링 방식이 증가함에 따라 REST API 방식이 자리 잡게 되었습니다.
그래서 `@Controller + @ResponseBody`를 함께 사용했어야 하는데, 두 가지를 함께 사용할 수 있도록 간소화된 어노테이션이 `@RestController`입니다.
따라서 `@RestController`를 사용한다면 다른 개발자에게 해당 컨트롤러는 뷰를 반환하지 않고 데이터만 반환하는 REST API 응답만 처리하는 컨트롤러라고 명시해줄 수 있습니다.

ddaaac commented on src/main/java/roomescape/controller/RoomescapeController.java:
```diff
@@ -0,0 +1,50 @@
+        return ResponseEntity.ok().body(ReservationResponseDto.toDto(reservation));
+    }
+
+    @DeleteMapping("/reservations/{id}")
+    public ResponseEntity<Void> deleteReservation(@PathVariable("id") Long id) {
+        try {
+            reservationRepository.deleteById(id);
+            return ResponseEntity.ok().build();
+        } catch (InvalidReservationException e){
+            return ResponseEntity.badRequest().build();
```
API를 자원이라는 관점에서 봤을 때 꼭 존재하지 않는 자원의 삭제에 대해서 꼭 예외를 던져야할까를 고민해보면 좋을 것 같아요.
"id에 해당하는 자원을 제거해줘"라는 응답에 이미 해당 자원이 없다면 그냥 200을 리턴해도 상관없지 않을까요?

ljhee92 replied:
범블비가 말씀하신 ‘API를 자원이라는 관점에서 본다’는 말이 어떤 의미인지 잘 이해되지 않아서 따로 검색을 좀 해보았어요!
`/reservations/{id}`와 같은 API는 '특정한 예약' 자원을 가리키는 것이고, 그 자원을 ‘삭제’하라는 요청인 것이니까 실제로 '삭제가 되었는 지 안 되었는 지'가 중요한 것이지 '삭제할 id가 존재하는 지'가 중요한 것이 아니라고 이해했어요.
제가 이해한 의미가 맞을까요?
그렇다면 badRequest를 반환하는 게 아니라 noContent를 반환하는 것이 더 이러한 관점에 맞다고도 생각되는데, 범블비는 어떻게 생각하시나요?

ddaaac replied:
204 안쓴지가 오래돼서 까먹고 있었는데... 잘 짚어주셨습니다!
204 쓰는게 더 적당해보여요 👍

ddaaac reviewed (changes requested):
안녕하세요 듀이 👋 

구현 깔끔하게 잘 해주셨네요.

간단한 코멘트 남겼으니 확인해주세요.

ljhee92 committed:
refactor: 클래스명 변경

- RoomescapeController -> ReservationController로 명시
- 클래스명에 Dto 삭제
- CreateReservationDto -> ReservationRequest로 변경
- 카멜케이스 적용

ljhee92 committed:
refactor: 웰컴 페이지 추가

- 웰컴 페이지는 정적 페이지이므로 static 하위에 위치

ljhee92 committed:
refactor: ViewController -> AdminViewController 클래스명 구체화, @RequestMapping 추가

- AdminViewController 내부에서 공유되는 mapping을 표현할 수 있도록 클래스 레벨에 @RequestMapping 추가

ljhee92 committed:
refactor: ReservationController에도 @RequestMapping 추가

ljhee92 committed:
refactor: 변수명 변경, 매직넘버 상수처리

ljhee92 committed:
refactor: 동시성 보장을 위해 List -> ConcurrentMap 변경

- AtomicLong을 사용한 이유인 동시성 보장과 동일한 이유로 List를 ConcurrentMap으로 변경

ljhee92 committed:
refactor: 매개변수 final 추가

ljhee92 committed:
refactor: 라인 컨벤션 통일

ddaaac reviewed (approved):
안녕하세요 듀이 👋 

질문 의도를 너무너무 잘 파악해주시고 꼼꼼하게 답변 남겨주셔서 할 말이 없네요 👍 

다음 단계 진행해주세요!
